File: src/codility/java/stack_queue/Fish.java
package codility.java.stack_queue;

import java.util.Stack;

public class Fish {

    public int solution(int[] A, int[] B) {
        // write your code in Java SE 8

        Stack<Integer> stack = new Stack<>();
        int ans = 0;
        for ( int i = A.length - 1; i >= 0; i--) {

            if (B[i] == 0) {
                stack.push(A[i]);
            } else {
                while (stack.size() > 0 && stack.peek() < A[i]) {
                    stack.pop();
                }

                if (stack.isEmpty()) {
                    ans++;
                }
            }
        }
        return ans + stack.size();
    }

}

File: src/codility/kotlin/Fish.kt
package codility.kotlin

import java.util.*

class Fish {

    fun solution(A: IntArray, B: IntArray): Int {
        val stack = Stack<Int>()

        var result = 0

        for (i in A.indices) {
            if (B[i] == 0) {

                while (stack.isNotEmpty() && stack.peek() < A[i]) {
                    stack.pop()
                }

                if (stack.isEmpty()) {
                    result++
                }

            } else {
                stack.push(A[i])
            }

        }

        return result + stack.size
    }

}